File: Python/23.merge-k-sorted-lists.py
#
# @lc app=leetcode id=23 lang=python
#
# [23] Merge k Sorted Lists
#
# https://leetcode.com/problems/merge-k-sorted-lists/description/
#
# algorithms
# Hard (36.73%)
# Likes:    3150
# Dislikes: 207
# Total Accepted:    480.9K
# Total Submissions: 1.3M
# Testcase Example:  '[[1,4,5],[1,3,4],[2,6]]'
#
# Merge k sorted linked lists and return it as one sorted list. Analyze and
# describe its complexity.
# 
# Example:
# 
# 
# Input:
# [
# 1->4->5,
# 1->3->4,
# 2->6
# ]
# Output: 1->1->2->3->4->4->5->6
# 
# 
#

# @lc code=start
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        # if len(lists) == 0:
        #     return None
        # if len(lists) == 1:
        #     return lists[0] 
        
        # l_merged = self.merge2Lists(lists[0], lists[1])
        # for i in range(2, len(lists)):
        #     l_merged = self.merge2Lists(l_merged, lists[i])

        # return l_merged

    #     if len(lists) == 0:
    #         return None
    #     interval = 1    
    #     while interval < len(lists):
    #         for i in range(0, len(lists)-interval, interval*2):
    #             lists[i] = self.merge2Lists(lists[i], lists[i+interval])
    #         interval *= 2
    #     return lists[0]

    # def merge2Lists(self, l1, l2):
    #     dummy = cur = ListNode(0)

    #     while l1 and l2: 
    #         if l1.val < l2.val:
    #             cur.next = l1 
    #             l1 = l1.next
    #         else:
    #             cur.next = l2 
    #             l2 = l2.next 
    #         cur = cur.next 
    #     cur.next = l1 or l2 
    #     return dummy.next

        # ln = []
        # for l in lists:
        #     while l:
        #         ln.append(l.val)
        #         l = l.next
        
        # ln.sort()

        # head = ListNode(0)
        # cur = head
        # for n in ln:
        #     cur.next = ListNode(n)
        #     cur = cur.next  
        # return head.next


        from heapq import heappush, heappop, heapreplace, heapify

        dummy = cur = ListNode(0)

        min_heap = [(node.val, node) for node in lists if node]
        heapify(min_heap)

        while min_heap:
            val, node = heappop(min_heap)
            if node.next:
                heappush(min_heap, (node.next.val, node.next))
            cur.next = node 
            cur = cur.next
        return dummy.next

# @lc code=end

File: C++/23.merge-k-sorted-lists.cpp
/*
 * @lc app=leetcode.cn id=23 lang=cpp
 *
 * [23] Merge k Sorted Lists
 */

#include<queue>
#include<tuple>
#include<vector>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// @lc code=start
class Solution {
public:

    // static bool cmp(ListNode* a, ListNode* b) {
    //         return a->val > b->val;
    // }

    // ListNode* mergeKLists(vector<ListNode*>& lists) {
    //     vector<ListNode*> queue;
    //     for(int i=0; i<lists.size(); i++){
    //         if(lists[i]) queue.push_back(lists[i]);
    //     }
    //     make_heap(queue.begin(), queue.end(), heapComp);


    //     ListNode dummy(0), *cur=&dummy;

    //     while(queue.size()>0) {
    //         cur->next = queue.front();

    //         pop_heap(queue.begin(), queue.end(), heapComp);
    //         queue.pop_back();

    //         cur = cur->next;
    //         if(cur->next) {
    //             queue.push_back(cur->next);
    //             push_heap(queue.begin(), queue.end(), heapComp);
    //         }
    //     }
    //     return dummy.next;

    struct heapComp {
        bool operator()(const ListNode* l, const ListNode* r) {
            return l->val > r->val;
        }
    };

    ListNode* mergeKLists(vector<ListNode*>& lists) {

        priority_queue<ListNode*, vector<ListNode*>, heapComp> queue;
        for (ListNode* node: lists){
            if(node) queue.push(node);
        }

        ListNode dummy(0), *cur = &dummy;

        while(!queue.empty()) {
            cur->next = queue.top();
            queue.pop();
            cur = cur->next;
            if(cur->next) queue.push(cur->next);
        }
        return dummy.next;
    }
};
// @lc code=end




class Solution {
public:

    vector<int> mergekSortedArrays(vector<vector<int>> &arrays) {
        
        struct heapComp {
            // to ensure a min-heap, otherwise it is max-heap by default
            bool operator() (tuple<int, int, int> &a, tuple<int, int, int> &b) {
                return get<2>(a) > get<2>(b);  
            }
        };

        tuple<int, int,  int>  tp;   //tuple<i,j,k> => k = arrays[i][j]
        priority_queue<tuple<int, int, int>, vector<tuple<int, int ,int>>, heapComp> pq;
        vector<int> result;
        
        int arrayLen = arrays.size();
        for (int i  = 0; i < arrayLen; ++i) {
            if (arrays[i].size() != 0) {
                pq.push(make_tuple(i, 0, arrays[i][0]));
            }
        }
        
        while(!pq.empty()) {
            tuple<int, int, int> elem = pq.top();
            pq.pop();
            result.push_back(get<2>(elem));
            get<1>(elem)++;
            if (get<1>(elem) < arrays[get<0>(elem)].size()) {
                pq.push(make_tuple(get<0>(elem),
                                   get<1>(elem),
                                   arrays[get<0>(elem)][get<1>(elem)]));
            }
            
        }
        
        return result;
    }
};